subsys/usb/device_next/class/usbd_uac2.c

```diff
@@ -0,0 +1,799 @@
+	IF_ENABLED(DT_NODE_HAS_COMPAT(node, zephyr_uac2_audio_streaming), (	\
+		COND_CODE_1(AS_HAS_EXPLICIT_FEEDBACK_ENDPOINT(node),		\
+			(UAC2_DESCRIPTOR_AS_FEEDBACK_EP_OFFSET(node),), (0,))	\
+	))
+#define DEFINE_AS_TERMINALS(node)						\
+	IF_ENABLED(DT_NODE_HAS_COMPAT(node, zephyr_uac2_audio_streaming), (	\
+		ENTITY_ID(DT_PROP(node, linked_terminal)),			\
+	))
+
+#define FREQUENCY_TABLE_NAME(node, i)						\
+	UTIL_CAT(frequencies_##i##_, ENTITY_ID(node))
+#define DEFINE_CLOCK_SOURCES(node, i)						\
+	IF_ENABLED(DT_NODE_HAS_COMPAT(node, zephyr_uac2_clock_source), (	\
+		static const uint32_t FREQUENCY_TABLE_NAME(node, i)[] =		\
+				DT_PROP(node, sampling_frequencies);		\
+	))
+
+#define DEFINE_LOOKUP_TABLES(i)							\
+	static const entity_type_t entity_types_##i[] = {			\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(i, DEFINE_ENTITY_TYPES)	\
+	};									\
+	static const uint16_t ep_offsets_##i[] = {				\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(i, DEFINE_AS_EP_OFFSETS)	\
+	};									\
+	static const uint16_t fb_offsets_##i[] = {				\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(i, DEFINE_AS_FB_OFFSETS)	\
+	};									\
+	static const uint8_t as_terminals_##i[] = {				\
+		DT_INST_FOREACH_CHILD_STATUS_OKAY(i, DEFINE_AS_TERMINALS)	\
+	};									\
+	DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(i, DEFINE_CLOCK_SOURCES, i)
+
+#define DEFINE_UAC2_CLASS_DATA(inst)						\
+	DT_INST_FOREACH_CHILD(inst, VALIDATE_NODE)				\
+	static struct uac2_ctx uac2_ctx_##inst;					\
+	static uint8_t uac2_descriptor_##inst[] = {				\
+		UAC2_DESCRIPTORS(DT_DRV_INST(inst))				\
+		0x00, 0x00 /* terminator required by USBD stack */		\
+	};									\
+	static struct usbd_class_data uac2_class_##inst = {			\
+		.desc = (struct usb_desc_header *)uac2_descriptor_##inst,	\
+		.priv = (void *)DEVICE_DT_GET(DT_DRV_INST(inst)),		\
+	};									\
+	USBD_DEFINE_CLASS(uac2_##inst, &uac2_api, &uac2_class_##inst);		\
+	DEFINE_LOOKUP_TABLES(inst)						\
+	static const struct uac2_cfg uac2_cfg_##inst = {			\
+		.node = &uac2_##inst,						\
+		.entity_types = entity_types_##inst,				\
+		.ep_offsets = ep_offsets_##inst,				\
+		.fb_offsets = fb_offsets_##inst,				\
+		.as_terminals = as_terminals_##inst,				\
+		.num_ifaces = ARRAY_SIZE(ep_offsets_##inst),			\
+		.num_entities = ARRAY_SIZE(entity_types_##inst),		\
+	};									\
+	BUILD_ASSERT(ARRAY_SIZE(ep_offsets_##inst) <= 32,			\
+		"UAC2 implementation supports up to 32 AS interfaces");		\
+	BUILD_ASSERT(ARRAY_SIZE(entity_types_##inst) <= 255,			\
+		"UAC2 supports up to 255 entities");				\
+	DEVICE_DT_DEFINE(DT_DRV_INST(inst), NULL, NULL,				\
+		&uac2_ctx_##inst, &uac2_cfg_##inst,				\
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,		\
+		NULL);
+DT_INST_FOREACH_STATUS_OKAY(DEFINE_UAC2_CLASS_DATA)
+
+static int clock_frequencies(struct usbd_class_node *const node, uint8_t id,
+			     const uint32_t **frequencies)
+{
+	const struct device *dev = node->data->priv;
+	int count;
+
+#define GET_FREQUENCY_TABLE(node, i)						\
```

Comment:
I don't quite agree, because the macro does use `} else` and looking at the code you wrote in the comment the first impression is that the `DT_INST_FOREACH_STATUS_OKAY(SELECT_FREQUENCY_TABLE)` is misaligned and won't ever execute because it is under `if (0)`.

Reply:
okay